### DISCUSSION TITLE
Rendering, Previewing & Baking

### DISCUSSION THREAD

**Discussion by muhuk** [muhuk/yer#5] [General]
This document tries to describe how the terrain data (inputs, artistic controls) is converted into output.

# Rendering Equation

We can think of rendering as a function:

```rust
fn sample(location: Vec2) -> PointSample { ... }
```

Where `PointSample` is some sorf of object that contains the height value for the `location` as well as other information like the biome, water level, curvature, ... (there can be user defined outputs)

Since our system will have multiple layers, we need to modify sample to be able to take advantage of previously calculated layers:

```rust
fn sample(base: PointSample, location: Vec2) -> PointSample
```

Then we can feed each layer with the previous layer's output and get an updated sample:

```rust
let mut point_sample = PointSample::default()
for layer in layers {
    point_sample = layer.sample(point_sample, location);
}
return point_sample;
```

Note that calling each layer (or operation) in correct order is important.

Both realtime preview and bake will use this same mechanism. The difference between the two will be sampling method and end result. Realtime preview needs to display a 3D mesh but bake may produce 2D images and other artifacts.

# Realtime Preview

Realtime preview is for getting an idea about the final result in a responsive way.

## Settings

**Boundaries:** Region to preview. (`(Vec2, Vec2)`)

**Max tessellation:** A power-of-two value that determines the preview mesh density when it is calculated all the way through.  Eg: a value of `512` will result in a grid mesh of 263169 (513 * 513) vertices and 524288 (512 * 512 * 2) triangles.

## Sampling Method

It is important for the preview to be as responsive as possible.  Therefore, to show some result as quickly as possible, a rough mesh should be calculated and, as long as the project is not edited in between, this mesh should be subdivided and only the new points are sampled.  As soon as the project is edited, preview mesh should be discarded and the process should start all over.

```
+-----+     +--o--+     +--+--+
-     -     -     -     -  -  - 
-     -     -     -     -  -  -
-     -  -> o  o  o  -> +--+--+
-     -     -     -     -  -  - 
-     -     -     -     -  -  -
+-----+     +--o--+     +--+--+
```

A naive implementation of this algorithm can calculate each iteration in one go. An optimized version can use a tiled approach.

Since edits invaludate the preview as a whole there is no need to syncronize with in-memory terrain data. Worst case an outdated preview can be displayed for a few frames.

Most important aspect of this implementation is; user should never feel the need to disable preview to be able to make edits.

# Baking

While the focus for preview was responsiveness, the focus for baking is quality.  So while baking the entire interface can be unrespondive and edits can be blocked.

Multiple baking targets can be set up. (Whereas there is only one preview setup)

## Bake Regions

Multiple regions (2D boundaries) can be set up in a project. It should be easy to set up multiple adjacent tiles.

## Bake Formats

For each bake setup one or more output formats can be specified.  Note that reusing samples might not be possible across different outputs.  For example; image based outputs would sample near the center of a pixel, but mesh outputs would sample near vertices.

For heightmaps (image), linear (EXR) and/or 16-bit options should be available.

## Supersampling

To improve output quality supersampling should be implemented. Basic (jittered) grid sampling (2x2, 4x4, 8x8, etc.) as well as monte carlo sampling (eg: 1000 samples but stop sampling when error rate drops below 0.01) should be available to user.

## Baking Pipeline

```
Terrain Data -> Sampler -> Gatherer -> Exporter
```

**Sampler** here is not different than the preview sampler.  **Gatherer** triggers sampling, then collects all the samples and stops sampling for a given pixel/vertex. **Exporter**s work at the level of individual artifacts (like a single image).  So we could have reversed the arrows in the diagram above.  However the samples should be re-used as much as possible (to avoid calculating the same thing twice) so ideally gatherer should be able to feed multiple exporters.  Hence the arrows indicate data flow direction.

**Comment by sniperliu** [comment]
Layer is 2D concept, I image the final terrain is projections of different 4D spaces.

**Reply by muhuk** [reply]
What do you mean by `4D spaces`? What is the 4th dimension?

Also zbrush has layers:

![image](https://github.com/user-attachments/assets/1910c396-9285-4da4-89fe-165c6bb2098c)

A layer is a conceptual unit of changes that are applied in a sequential manner.
